sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java

```diff
@@ -0,0 +1,575 @@
+
+      //get published segments from coordinator
+      final JsonParserIterator<DataSegment> metadataSegments = getMetadataSegments(
+          druidLeaderClient,
+          jsonMapper
+      );
+
+      Set<String> availableSegmentIds = new HashSet<>();
+      final FluentIterable<Object[]> availableSegments = FluentIterable
+          .from(() -> availableSegmentEntries)
+          .transform(val -> {
+            try {
+              if (!availableSegmentIds.contains(val.getKey().getIdentifier())) {
+                availableSegmentIds.add(val.getKey().getIdentifier());
+              }
+              return new Object[]{
+                  val.getKey().getIdentifier(),
+                  val.getKey().getDataSource(),
+                  val.getKey().getInterval().getStart(),
+                  val.getKey().getInterval().getEnd(),
+                  val.getKey().getSize(),
+                  val.getKey().getVersion(),
+                  val.getKey().getShardSpec().getPartitionNum(),
+                  val.getValue().getNumReplicas(),
+                  val.getValue().getNumRows(),
+                  val.getValue().isPublished(),
+                  val.getValue().isAvailable(),
+                  val.getValue().isRealtime(),
+                  jsonMapper.writeValueAsString(val.getKey())
+              };
+            }
+            catch (JsonProcessingException e) {
+              log.error(e, "Error getting segment payload for segment %s", val.getKey().getIdentifier());
+              throw new RuntimeException(e);
+            }
+          });
+
+      final FluentIterable<Object[]> publishedSegments = FluentIterable
+          .from(() -> metadataSegments)
+          .transform(val -> {
+            try {
+              if (availableSegmentIds.contains(val.getIdentifier())) {
+                return null;
+              }
+              return new Object[]{
+                  val.getIdentifier(),
+                  val.getDataSource(),
+                  val.getInterval().getStart(),
+                  val.getInterval().getEnd(),
+                  val.getSize(),
+                  val.getVersion(),
+                  val.getShardSpec().getPartitionNum(),
+                  0,
+                  -1,
+                  1,
+                  0,
+                  0,
+                  jsonMapper.writeValueAsString(val)
+              };
+            }
+            catch (JsonProcessingException e) {
+              log.error(e, "Error getting segment payload for segment %s", val.getIdentifier());
+              throw new RuntimeException(e);
+            }
+          });
+
+      Iterable<Object[]> allSegments = Iterables.unmodifiableIterable(
+          Iterables.concat(availableSegments, publishedSegments));
+
+      return Linq4j.asEnumerable(allSegments).where(t -> t != null);
+
+    }
+
+    // Note that coordinator must be up to get segments
+    JsonParserIterator<DataSegment> getMetadataSegments(
```

Comment:
Looks like this can be a static method outside of this class

Reply:
yes, made private static